### DISCUSSION TITLE
A question regarding spy and CaughtException

### DISCUSSION THREAD

**Discussion by kubk** [mobxjs/mobx#3341] [Q&A]
If you run [this codesandbox](https://codesandbox.io/s/tender-villani-8hc6l1) you'll notice that there is the following `spy` report: 

```
{
observableKind: "computed"
debugObjectName: "ObservableObject@1.isEven"
object: Object
type: "update"
oldValue: CaughtException
newValue: true
}
```

Notice the `oldValue: CaughtException`. Technically the spy report is correct because the initial value of a computed is `CaughtException`: https://github.com/mobxjs/mobx/blob/e60b36c9c78ff9871be1bd324831343c279dd69f/packages/mobx/src/core/computedvalue.ts#L91

But from a library consumer point of view the initial value of `isEven` computed is `true`. I've noticed it while developing a logging library for Mobx. I keep getting log messages like this:
<img width="449" alt="Screen Shot 2022-03-19 at 10 53 58" src="https://user-images.githubusercontent.com/22447849/159112878-05cd29c1-ecfa-4634-b912-fcd23d7b4179.png">

I'd like to filter such messages out. But neither `CaughtException` nor `isCaughtException` is a part of public Mobx API. What'd you suggest me to solve the issue?

**Comment by urugator** [comment]
If we export `isCaughtException` won't there be some problem with differentiating initial value and actual error?
Do you need `isCaughtException` in other places?

Btw ES newly supports `cause` on standard `Error` maybe it would be nice to make use of it.

**Reply by kubk** [reply]
@urugator Thanks, I decided to use `value?.constructor?.name === 'CaughtException'` check to determine if value is `CaughtException`. So there is no need to import `isCaughtException`. 

> won't there be some problem with differentiating initial value and actual error?

It looks I don't need to differentiate them. I'll just hide all the logs related to `CaughtException`

**Reply by urugator** [reply]
Occured to me as well, but then I thought it would not work when minified.... ?

**Comment by mweststrate** [comment]
Yeah depends on minifier settings, some configs will minify the constructor
name.

**Comment by kubk** [comment]
@mweststrate @urugator Thank you, you were right. I've built `mobx` locally and see that `CaughtException` doesn't exist in the prod build, because the property mangling is enabled:

https://github.com/mobxjs/mobx/blob/56bfb6796b4462ce056109e02c63697f0d21e4fe/packages/mobx/tsdx.config.js#L25

> Btw ES newly supports cause on standard Error maybe it would be nice to make use of it.

Wouldn't it increase the bundle size? Now it makes sense for me to make `isCaughtException` a part of public Mobx API, it won't affect the bundle size. I'd be happy to hear your opinions.

**Reply by urugator** [reply]
It's not because of property mangling (only props ending with _ are mangled), simply the function name is minified.

Unless there are some problems with doing so, I propose to:
Replace
https://github.com/mobxjs/mobx/blob/2caf7e1a3504dde3d7c9bde3c6fb56ca85168018/packages/mobx/src/core/derivation.ts#L63-L67
with
```javascript
class DerivationError extends Error {};
```
Remove `isCaughtException` and replace the calls with `instanceof DerivationError`.
Replace `new CaughtException(cuase)` with `new DerivationError("Something like: Error from derivation named ...", { cause })`
Export `DerivationError` as public API
Replace computed's initial value with `Symbol('initialComputedValue')` (so we don't have to create heavy weight error object for each computed, also I think we don't need unique initial value for each computed) and export as `_computedInitialValue` or alternatively `_isComputedInitialValue`.

**Reply by kubk** [reply]
@urugator That is a lot of work 😃 I think if we change the initial value of computed to `Symbol` it'd solve the initial issue. I'd skip logging `Symbol`s because they can't be observable. Do you think changing `CaughtException` to `Symbol('initialValue')` may bring some performance optimizations?

**Reply by urugator** [reply]
> I'd skip logging Symbols because they can't be observable.

I don't follow. `oldValue` can be whatever returned from `computed` ... the current `CaughtException` isn't observable as well.

> Do you think changing CaughtException to Symbol('initialValue') may bring some performance optimizations?

In principle yes, in practice most likely no. Not sure it there is some technical reason for initial values being unique per computed instance. Failing tests should tell if that's the case I guess.

I am fine with changing initial value to symbol and exporting it, I just don't want to expose CaughtException in it's current form.

**Reply by urugator** [reply]
Just occured to me maybe you could also solve with with a little "hack". On the first computed spy event, capture the initial value somewhere and then for all other computed spy events:
`Object.getPrototypeOf(initialValue) === Object.getPrototypeOf(oldValue)) && oldValue.cause === null`. You can create a dummy computed to force the first spy event with oldValue being CaughtException.

**Reply by mweststrate** [reply]
Symbols are definitely better than Exceptions here indeed (blame me). If symbols cause trouble, you could also reserve a constant unique, frozen object for it, something like `const UninitializedMarker = { MobxUninitializedValue: true }`.

**Reply by mweststrate** [reply]
It'd be fine to expose that object as well, as long as clearly named and sealed, to make it possible to reflect on it.
